tests/http_spec.rs

```diff
@@ -0,0 +1,215 @@
+    pub body: Option<serde_json::Value>,
+  }
+  #[derive(Deserialize, Clone, Debug)]
+  pub struct APIResponse {
+    pub status: u16,
+    pub headers: Option<BTreeMap<String, String>>,
+    pub body: Option<serde_json::Value>,
+  }
+  #[derive(Deserialize, Clone, Debug)]
+  pub struct UpstreamRequest(pub APIRequest);
+  #[derive(Deserialize, Clone, Debug)]
+  pub struct UpstreamResponse(APIResponse);
+  #[derive(Deserialize, Clone)]
+  pub struct DownstreamRequest(pub APIRequest);
+  #[derive(Deserialize, Clone)]
+  pub struct DownstreamResponse(pub APIResponse);
+  #[derive(Deserialize, Clone)]
+  pub struct DownstreamAssertion {
+    pub request: DownstreamRequest,
+    pub response: DownstreamResponse,
+    pub annotation: Option<Annotation>,
+  }
+
+  #[derive(Default, Deserialize, Clone)]
+  pub struct HttpSpec {
+    pub config: String,
+    pub name: String,
+    pub description: Option<String>,
+    pub upstream_mocks: Vec<(UpstreamRequest, UpstreamResponse)>,
+    pub expected_upstream_requests: Vec<UpstreamRequest>,
+    pub downstream_assertions: Vec<DownstreamAssertion>,
+  }
+
+  impl HttpSpec {
+    fn read(spec: &str) -> Option<Self> {
+      spec
+        .split('.')
+        .last()
+        .and_then(|ext| match ext.to_lowercase().as_str() {
+          "json" => Self::read_json(spec),
+          "yaml" => Self::read_yaml(spec),
+          _ => None,
+        })
+    }
+    fn read_json(spec: &str) -> Option<Self> {
+      let contents = fs::read_to_string(spec).ok()?;
+      let spec = serde_json::from_str(&contents);
+
+      spec.ok()
+    }
+
+    fn read_yaml(spec: &str) -> Option<Self> {
+      let contents = fs::read_to_string(spec).ok()?;
+      let spec = serde_yaml::from_str(&contents);
+      spec.ok()
+    }
+    async fn setup(&self) -> Arc<ServerContext> {
+      let config = Config::from_file_paths([self.config.clone()].iter())
+        .await
+        .ok()
+        .unwrap();
+      let blueprint = Blueprint::try_from(&config).unwrap();
+      let client = Arc::new(MockHttpClient { upstream_mocks: Arc::new(self.upstream_mocks.to_vec()) });
+      let server_context = ServerContext::new(blueprint, client);
+      Arc::new(server_context)
+    }
+  }
+
+  #[derive(Clone)]
+  struct MockHttpClient {
+    upstream_mocks: Arc<Vec<(UpstreamRequest, UpstreamResponse)>>,
+  }
+  #[async_trait::async_trait]
+  impl HttpClient for MockHttpClient {
+    async fn execute(&self, req: reqwest::Request) -> anyhow::Result<Response> {
+      let upstream_mocks = self.upstream_mocks.clone();
+      let upstream_mock = upstream_mocks.iter().find(|(upstream_request, _res)| {
+        let method_match = req.method().as_str()
+          == serde_json::to_string(&upstream_request.0.method.clone().unwrap_or_default())
+            .unwrap()
+            .as_str()
+            .trim_matches('"');
+        let url_match = req.url().as_str() == req.uri().to_string().as_str();
+        method_match && url_match
```

Comment:
splitting the big `execute` function into smaller functions could significantly improve readability of the code

Reply:
Http_client trait ony has execute fn defined defining the function outside of trait is not a good idea, added comments for readability